src/experimental/wildcard.ts

```diff
@@ -0,0 +1,104 @@
+import fs from 'fs/promises'
+import type { Dirent } from 'fs'
+import path from 'path'
+import { SRC } from '../constants'
+import { ExportCondition } from '../types'
+import {
+  filenameWithoutExtension,
+  hasAvailableExtension,
+  nonNullable,
+} from '../utils'
+
+// TODO: support nested wildcard exports (e.g. './foo/*')
+const getWildcardExports = (
+  exports: Record<string, ExportCondition>,
+): Record<string, ExportCondition> => {
+  return { './*': exports['./*'] }
+}
+
+const isExportable = async (
+  dirent: Dirent,
+  pathname: string,
+): Promise<boolean> => {
+  if (dirent.isDirectory()) {
+    const innerDirents = await fs.readdir(path.join(pathname, dirent.name), {
+      withFileTypes: true,
+    })
+    return innerDirents.some(
+      ({ name }) => name.startsWith('index') && hasAvailableExtension(name),
+    )
+  }
+  return (
+    dirent.isFile() &&
+    !dirent.name.startsWith('index') &&
+    hasAvailableExtension(dirent.name)
+  )
+}
+
+async function getExportables(cwd: string): Promise<string[]> {
+  const pathname = path.resolve(cwd, SRC)
+  const dirents = await fs.readdir(pathname, { withFileTypes: true })
+  const exportables: (string | undefined)[] = await Promise.all(
+    dirents.map(async (dirent) =>
+      (await isExportable(dirent, pathname)) ? dirent.name : undefined,
+    ),
+  )
+  return exportables.filter(nonNullable)
+}
+
+function mapWildcard(
+  wildcardExports: Record<string, ExportCondition>,
+  exportables: string[],
+): ExportCondition[] {
+  return exportables.map((exportable) => {
+    const filename = exportable.includes('.')
+      ? filenameWithoutExtension(exportable)
+      : undefined
+
+    if (!filename) {
+      return {
+        [`./${exportable}`]: JSON.parse(
+          JSON.stringify(wildcardExports['./*']).replace(
+            /\*/g,
+            `${exportable}/index`,
+          ),
+        ),
+      }
+    }
+    return JSON.parse(JSON.stringify(wildcardExports).replace(/\*/g, filename))
+  })
+}
+
+export async function resolveWildcardExports(
+  exports: string | Record<string, ExportCondition> | undefined,
+  cwd: string,
+): Promise<ExportCondition | undefined> {
+  if (!exports || typeof exports === 'string') return undefined
+  const hasWildcard = !!exports['./*']
+
+  if (hasWildcard) {
+    console.warn(
+      `The wildcard export "./*" is experimental and may change or be removed at any time.\n` +
+        'To open an issue, please visit https://github.com/huozhi/bunchee/issues' +
+        '.\n',
+    )
+  } else {
+    return undefined
+  }
+
+  const exportables = await getExportables(cwd)
+  if (exportables.length > 0) {
+    const wildcardExports = getWildcardExports(exports)
+    const resolvedWildcardExports = mapWildcard(wildcardExports, exportables)
+
+    const resolvedExports = Object.assign(
+      {},
+      exports,
+      ...resolvedWildcardExports,
+      exports,
```

Comment:
not sure why we need to merge exports twice here?

Reply:
To preserve the given export keys by overwriting...

```ts
{
  "name": "my-lib",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js"
    },
    "./server": {
      "types": "./dist/server.d.ts",
      "require": "./dist/server.cjs"
    },
    "./*": {
      "types": "./dist/*.d.ts",
      "import": "./dist/*.js"
    }
  }
}

// converted as

{
  "name": "my-lib",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js"
    },
    "./server": {
      "types": "./dist/server.d.ts",
      "require": "./dist/server.cjs"
    },
    "./a": {
      "types": "./dist/a/index.d.ts",
      "import": "./dist/a/index.js"
    }
    "./b": {
      "types": "./dist/b.d.ts",
      "import": "./dist/b.js"
    }
  }
}
```

Or I could just manage them with loops, but was wondering if it could be shorter.

Reply:
I think you can clone the `exportsCondition` first `{...exportsCondition }`, get the mapped wildcad ones, then delete the whildcard from the cloned exportsCondition, then merge them.

Reply:
That was my initial work, but since `server` is also exportable, I had to read the existing keys to filter them out before mapping the wildcards.

Which is reasonable, but wanted to reduce codes as much as possible and was wondering if there is a simpler process.

I'll fix it, thank you!